library/ssl_msg.c

```diff
@@ -574,6 +578,33 @@ int mbedtls_ssl_encrypt_buf( mbedtls_ssl_context *ssl,
         return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
     }
 
+    /* The following two code paths implement the (D)TLSInnerPlaintext
+     * structure present in TLS 1.3 and DTLS 1.2 + CID.
+     *
+     * See ssl_build_inner_plaintext() for more information.
+     *
+     * Note that this changes `rec->data_len`, and hence
+     * `post_avail` needs to be recalculated afterwards.
+     *
+     * Note also that the two code paths cannot occur simultaneously
+     * since they apply to different versions of the protocol. There
+     * is hence no risk of double-addition of the inner plaintext.
+     */
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
+    if( transform->minor_ver == MBEDTLS_SSL_MINOR_VERSION_4 )
+    {
+        if( ssl_build_inner_plaintext( data,
+                        &rec->data_len,
+                        post_avail,
+                        rec->type ) != 0 )
+        {
+            return( MBEDTLS_ERR_SSL_BUFFER_TOO_SMALL );
+        }
+
+        rec->type = MBEDTLS_SSL_MSG_APPLICATION_DATA;
```

Comment:
@yanesca Yes, with TLS 1.3, we have to separate wire-version from logical version since TLS 1.3 records use the TLS 1.2 version field on the wire. In fact, they're indistinguishable from TLS 1.2 records.

There's no single solution in how this should be reflected in the values of `rec->version`. There are at least the following options:
* Always use the on-the-wire version byte for `rec->version`, and don't modify it within the record protection routines. This will require some logical-to-wire version conversion setting up an outgoing record prior to encryption. For incoming records, one could in theory correct the version byte `rec->version` after record decryption, but it's of no use since the real protocol version is always taken from `transform->minor_ver`. 
* Switch to/from wire/logical version for `rec->version` in the record protection routines. This would strengthen the encapsulation slightly, but on the other hand, it introduces code elsewhere since in contrast to the content type, there's no hiding of the real version that's specific to encrypted records: Even the initial ClientHello / ServerHello wouldn't use a TLS 1.3 version identifier in the record header. So, if we'd set `rec->version` to be the logical version prior to calling `ssl_encrypt_buf()` - which so far only happens for non-NULL transforms - then we'd need to change something to make sure we don't write the wrong version to the wire.

My inclination is to go with the first option, and to use `rec->version` as a functionally irrelevant copy of the wire-version. In this case, there's nothing to necessarily to change in this PR, because `rec->version` can be anything - it has no impact on record protection. Only when we start using the TLS 1.3 protection from the SSL layer, we have to set it correctly to get the right wire-format.

Reply:
I agree that conceptually it is best to cleanly separate the logical and the wire version. I wasn't looking for a final solution here, for something correct, that we can refactor later if needed. 

If we are willing to put more thought into it, then I think I would prefer to do it in the conversion functions. Converting between them is straightforward and is done by `mbedtls_ssl_read/write_version()`. The problem is that in TLS 1.3, we have two kinds of wire versions: `version` and `legacy_version`. To resolve this we could have separate functions for each. Since `version` is used only in the supported versions extension, it could be safe to change `mbedtls_ssl_write_version()` to write 0x0303 for TLS 1.3 too. (`mbedtls_ssl_read_version()` does not need to change, because in TLS 1.3 the legacy version must be ignored.)

Of course we will need to add a second pair of functions for this conversion when we are writing the supported versions extension. We should document the functions to make this job for our future selves easier.

What do you think?

Reply:
Yes, I agree, and I've got something along those lines in preparation, perhaps it could be one of the next chunks to upstream. However, I don't think it belongs to this PR, because it's about writing/reading to/from the wire, while the PR is about the record conversion to/from unprotected/protected.

Reply:
Ok, let's do it in one of the next chunks then.